Gdnative windows interface to obtain GL context info

BastiaanOlij opened the pull request:
This PR adds two access methods to GDNative that give GDNative modules access to the GL context created on Windows. We need access to these to make our new OpenXR module work on Windows.

BastiaanOlij committed:
Adding access methods for Windows OS for GDNative

akien-mga commented on platform/windows/os_windows.cpp:
```diff
@@ -1852,6 +1852,11 @@ void OS_Windows::finalize_core() {
+ContextGL_Windows *OS_Windows::get_gl_context() {
+	return gl_context;
+}
+
+
```
Style issue :)

BastiaanOlij commented:
@akien-mga ok I've confirmed this now works properly, we need this logic in order to make OpenXR work but it's a bit of an issue that in Godot 4 this will need to work differently or just return proper results  if the OpenGLES2 driver is used seeing these functions have no meaning under Vulkan. I think for now in master we should just make sure the API entry points are there but they just return 0 for now.

bruvzg commented:
Window handles also can be quite handy to have (specially for X11), maybe we should have universal API for all kinds of native handles on all platforms instead of making extensions for each one, something like:

```c++
enum Handle {
	APPLICATION_HANDLE, //HINSTANCE, NSApplication*, UIApplication*, JNIEnv* ...
	DISPLAY_HANDLE, // X11::Display* ...
	WINDOW_HANDLE, // HWND, X11::Window*, NSWindow*, UIWindow*, Android activity ...
	WINDOW_VIEW, // HDC, NSView*, UIView*, Android surface ...
	OPENGL_INSTANCE, // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
	VULKAN_INSTANCE, // VkInstance
	VULKAN_SURFACE, // VkSurfaceKHR
	VULKAN_DEVICE, // VkDevice
	...
};

void *godot_get_native_handle(Handle p_handle_type, int p_index);
```

And it's gonna be easier to extend without breaking compatibility.

BastiaanOlij commented:
@bruvzg I like that idea, looking at us having this as an API on Android already this would be a nicer solution.
I'll work on this over the weekend and change this PR to follow that approach.

@m4gr3d what is your take on this? could this replace the custom API we added for android or where there other things that we added? If we can replace it into a single system we can deprecate the android one for 4.

m4gr3d commented:
@bruvzg @BastiaanOlij I'm not sure I follow. Can you provide an example how the suggested api would be used?

BastiaanOlij commented:
@m4gr3d 

instead of calling `android_api->godot_android_get_env()` you would call `api->godot_get_native_handle(APPLICATION_HANDLE, 0)`

It means we can just add more enums as we expose more internal values instead of having to add new API calls or platform specific API calls. 

@bruvzg what is the index for?

BastiaanOlij commented:
I don't like the way currently we have the next->next->next stuff that would add structs with one or two new APIs though. I wonder if there is a better way to organise those

bruvzg commented:
> what is the index for?

For handling multiple windows.

m4gr3d commented:
> @m4gr3d
> 
> instead of calling `android_api->godot_android_get_env()` you would call `api->godot_get_native_handle(APPLICATION_HANDLE, 0)`
> 
> It means we can just add more enums as we expose more internal values instead of having to add new API calls or platform specific API calls.

@BastiaanOlij yes that sounds good, thanks for the clarification!

BastiaanOlij commented:
@bruvzg ah ok 👍 

I was thinking, wouldn't it make sense to add get_native_handle as a function to the OS class and expose it in that way? Then we don't need to change gdnative at all, it just becomes a new method we can call through normal bindings?

bruvzg commented:
> I was thinking, wouldn't it make sense to add get_native_handle as a function to the OS class and expose it in that way?

Not sure if `void *` functions are supported by `ClassDB` and can be exposed, but if it, this is even better solution.

BastiaanOlij commented:
> > I was thinking, wouldn't it make sense to add get_native_handle as a function to the OS class and expose it in that way?
> Not sure if `void *` functions are supported by `ClassDB` and can be exposed, but if it, this is even better solution.
I'll have to play with it, in the end void * is an integer so in theory we can return it as that.

BastiaanOlij commented:
This will likely be replaced by #42531
